overlord/devicestate/systems.go

```diff
@@ -176,3 +186,212 @@ func seededSystemFromModeenv() (*seededSystem, error) {
+			return nil, "", err
+		}
+	}
+	if err := w.SetOptionsSnaps(optsSnaps); err != nil {
+		return nil, "", err
+	}
+
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		fromDB := func(ref *asserts.Ref) (asserts.Assertion, error) {
+			return ref.Resolve(db.Find)
+		}
+		return asserts.NewFetcher(db, fromDB, save)
+	}
+	f, err := w.Start(db, newFetcher)
+	if err != nil {
+		return nil, "", err
+	}
+	// past this point the system directory is present
+
+	localSnaps, err := w.LocalSnaps()
+	if err != nil {
+		return nil, recoverySystemDir, err
+	}
+
+	for _, sn := range localSnaps {
+		info, ok := modelSnaps[sn.Path]
+		if !ok {
+			return nil, recoverySystemDir, fmt.Errorf("internal error: no snap info for %q", sn.Path)
+		}
+		// TODO: the side info derived here can be different from what
+		// we have in snap.Info, but getting it this way can be
+		// expensive as we need to compute the hash, try to find a
+		// better way
+		_, aRefs, err := seedwriter.DeriveSideInfo(sn.Path, f, db)
+		if err != nil {
+			if !asserts.IsNotFound(err) {
+				return nil, recoverySystemDir, err
+			} else if info.SnapID != "" {
+				// snap info from state must have come
+				// from the store, so it is unexpected
+				// if no assertions for it were found
+				return nil, recoverySystemDir, fmt.Errorf("internal error: no assertions for asserted snap with ID: %v", info.SnapID)
+			}
+		}
+		if err := w.SetInfo(sn, info); err != nil {
+			return nil, recoverySystemDir, err
+		}
+		sn.ARefs = aRefs
+	}
+
+	if err := w.InfoDerived(); err != nil {
+		return nil, recoverySystemDir, err
+	}
+
+	for {
+		// get the list of snaps we need in this iteration
+		toDownload, err := w.SnapsToDownload()
+		if err != nil {
+			return nil, recoverySystemDir, err
+		}
+		// which should be empty as all snaps should be accounted for
+		// already
+		if len(toDownload) > 0 {
+			which := make([]string, 0, len(toDownload))
+			for _, sn := range toDownload {
+				which = append(which, sn.SnapName())
+			}
+			return nil, recoverySystemDir, fmt.Errorf("internal error: need to download snaps: %v", strings.Join(which, ", "))
+		}
+
+		complete, err := w.Downloaded()
+		if err != nil {
+			return nil, recoverySystemDir, err
+		}
+		if complete {
+			logger.Noticef("snap processing complete")
+			break
+		}
+	}
+
+	for _, warn := range w.Warnings() {
+		logger.Noticef("WARNING: %s", warn)
+	}
+
+	unassertedSnaps, err := w.UnassertedSnaps()
+	if err != nil {
+		return nil, recoverySystemDir, err
+	}
+	if len(unassertedSnaps) > 0 {
+		locals := make([]string, len(unassertedSnaps))
+		for i, sn := range unassertedSnaps {
+			locals[i] = sn.SnapName()
+		}
+		logger.Noticef("system %q contains unasserted snaps %s", label, strutil.Quoted(locals))
+	}
+
+	copySnap := func(name, src, dst string) error {
+		if osutil.FileExists(dst) && strings.HasPrefix(dst, assertedSnapsDir+"/") {
+			// unasserted snaps are not shared
```

Comment:
this comment is strange  because at this point we are in the shared case, no?  maybe it needs to talk about the if condition positively or be expanded?

Reply:
yeah this is confusing to me as well, I am reading this as "if the destination is in the asserted snaps directory the snap is an asserted snap and as such doesn't need to be copied since asserted snap revisions will always be the same file content" - which I don't yet know if that is true or not from reading the code, but that does not agree with the comment inside the if condition, I would expect this whole block to be something more like:

```go
// if the destination snap is in the asserted snaps dir and already exists, we don't need to 
// copy it since asserted snaps are shared
if strings.HasPrefix(dst, assertedSnapsDir+"/") && osutil.FileExists(dst) {
    return nil
}
// otherwise, unasserted snaps are not shared, so even if the destination
// already exists if it is not in the asserted snaps we should copy it
}
```

or something else which is not as verbose :smiley:

Reply:
I used my own suggestion here :-D feel free to refine in a followup